File: Cpp/826_Most_Profit_Assigning_Work.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        int result = 0;
        priority_queue<pair<int,int>> pq;
        for(int i=0; i<profit.size(); ++i) {
            pq.push({profit[i], difficulty[i]});
        }
        
        sort(worker.begin(), worker.end(), greater<int>());
        
        for(int i=0; i<worker.size(); ++i) {
            while(!pq.empty() && pq.top().second > worker[i]) {
                pq.pop();
            }
            if(!pq.empty()) {
                result+= pq.top().first;
            }
        }
        
        return result;
    }
};

File: Kotlin/826_Most_Profit_Assigning_Work.kt
class Solution {
    fun maxProfitAssignment(difficulty: IntArray, profit: IntArray, worker: IntArray): Int {
        var result = 0;
        val pq = PriorityQueue(Comparator<Job>{ a, b ->
            if (a.profit == b.profit) a.difficulty - b.difficulty else b.profit - a.profit
        })
        
        for(i in profit.indices) {
            pq.offer(Job(difficulty[i], profit[i]))
        }
        
        worker.sortDescending()

        worker.forEach {
            while(!pq.isEmpty() && pq.peek().difficulty > it) {
                pq.poll()
            }
            if(!pq.isEmpty()) {
                result += pq.peek().profit
            }
        }
        
        return result
    }
}

data class Job(val difficulty: Int, val profit: Int)